### ISSUE TITLE
precise weapon placement

### ISSUE THREAD

**Issue by ohrrpgce-bugbot** [ohrrpgce/ohrrpgce#58]
[bz#58]

A longtime complaint is the difficulty (impossibility) of accurately aligning a
hero's hand with his or her weapon. An alignment table for weapons would not be
so hard to implement, and would really be worth it.

This is a priority for the Rusalka release

From: @bob-the-hamster
Severity:     feature request
Depends on:   bz#30 (#1043)

**Comment by ohrrpgce-bugbot** [comment]
Comment author: **Yuriy Melnichenko &lt;yuriym19&gt;**

hm... do weapons have their own lump?  if so than its definately won't be hard 
to implement, if yes, than we probably need to do that.
I have an idea for this, not sure if its a good one though.  you know that you 
can place enemies in the formation, right?  why not do same principle for 
weapon placement?

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

Weapons pictures are in the .PT6 lump, but there is no separate lump for weapon
data.

The hero data lump .DT0 will need to make room for an X and Y "hand position"
and The weapon's "handle position" should probably go in the .ITM lump where
other weapon data is stored.

I would like the weapon alignment editor to be a part of the hero editor. In the
weapon alignment editor, you will be able to position any hero's hand, and any
weapon's handle, and preview a combination of any hero and weapon

Although the principle of weapon alignment is a little bit like enemy alignment
in formations, it is not similar enough to warrant doing them the same way.
(although speaking of that, hero positioning in battle might be a worthy feature)

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @pkmnfrk

\*\*\* Bug bz#29 (#1042) has been marked as a duplicate of this bug. \*\*\*

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @raekuul


> Weapons pictures are in the .PT6 lump, but there is no separate lump for 
weapon
> data.
> The hero data lump .DT0 will need to make room for an X and Y "hand position"
> and The weapon's "handle position" should probably go in the .ITM lump where
> other weapon data is stored.
> I would like the weapon alignment editor to be a part of the hero editor. In 
the
> weapon alignment editor, you will be able to position any hero's hand, and any
> weapon's handle, and preview a combination of any hero and weapon
> Although the principle of weapon alignment is a little bit like enemy 
alignment
> in formations, it is not similar enough to warrant doing them the same way.
> (although speaking of that, hero positioning in battle might be a worthy 
feature)

How would the UI for that work? Would we have to punch in Coordinates, or would 
we have to set the hand/handle points in a Graphic-Editor-esque thingy?

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

This will be implemented with two new editors, a "Handle Editor" inside the item
editor, and a "Hand Editor" inside the hero editor. This will not be a matter of
typing in X/Y coordinates. You will actually see the weapon or hero, and
position a flashing dot over the point at which hand and weapon will be joined.

The battle system will handle lining these points up when it does animations.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @pkmnfrk

I'm thinking that we should start off with just typing in the coordinates. I
don't know if we can fit another version of the graphics editor in there...

However, to compliment this, I think the existing graphics editor should display
the X,Y under the mouse (unless it already does this and I'm blind (well, I AM
blind (glasses), but...))

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

No worries. This will actually be a fairly simple editor. CUSTOM is not nearly
as tight on memory as GAME is, so there will be plenty of room for it.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @pkmnfrk

\*\*\* Bug bz⁠#115 has been marked as a duplicate of this bug. \*\*\*

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @pkmnfrk

Ok, completely ignoring the previous commentary, here's what I've done:

One thing we neglected to remember was that weapons and heros both have two
related frames, so they each need two positions.

So, in the Item Editor, there is now two new menu items ("handle X" and "handle
Y"). However, there is also a little cursor on the picture that is updated in
real time.

To allow editing the position for both frames, I made it so hitting "," (which
is "<") will flip to frame 0, and "." (">") flips to frame 1. The menu items are
updated with the related coordinates, and the cursor is updated as well.

In the hero editor, a similar scheme is used. However, the related options are
under Appearance and Misc. In that menu, the animated hero pic freezes on Attack
 A (and can be switched to B in the same fashion as the Item editor), and the
cursor shows up. (Oh, and there's the Hand X and Hand Y options).

Finally, in the hero animation sub (which positions the weapon graphics), it
reads in the appropriate values from the lumps, and positions the weapons
(hopefully properly).

I arbitrarily decided to allow a range of +/- 100 for both coordinates, so
people have plenty of wiggle room. In fact, they could do something like having
a boomerang go half-way across the screen if they felt like it. :)

I've tested it, and it seems to work pretty good. Someone else test it, please.
I won't mark this fixed until someone else tests it (or we gear up to release
Tirgoviste, which ever happens first <_<)

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

As far as I can tell, this is fixed \*except\* for the problem described in bug bz#30 (#1043).

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

This feature is reported to not work by multiple people using the latest nightlies

http://castleparadox.com/ohr/viewtopic.php?t=4501

And I was thinking, although this is a user-interface thing, and unrelated to the above-mentioned problem, rather than inputting an X and y value, we should just have a menu item that says "Set weapon placement" and when you select it, you see nothing but the hero (or weapon) and the hotspot cursor, which you position with the arrow keys, and press enter or space when you are done.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @arperry

Isn't this finished? Marking worksforme.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @pkmnfrk

No, it's not. I can guarantee that this feature doesn't work properly.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @pkmnfrk

BAM!

Hear that? It's the last nail in the coffin of this bug. All's peachy now.
